citar-file.el

```diff
@@ -112,12 +112,15 @@ Example: ':/path/to/test.pdf:PDF'."
              (extension)
              (seq-map
               (lambda (directory)
-                (expand-file-name
-                 (concat key "." extension) directory))
+                (directory-files directory t
+                                         (s-concat "^" (regexp-quote key)
+                                                   ".*\\("
+                                                   extension
+                                                   "\\)$")))
               dirs)))
-    (let* ((results-key (seq-mapcat
-                         #'possible-file-names-with-extension
-                         extensions))
+    (let* ((results-key (apply #'append (seq-mapcat
+                                         #'possible-file-names-with-extension
+                                         extensions)))
```

Comment:
Maybe instead of making a regexp per extension and looping, you could make a single regexp that matches all the extensions? Something like this:
```suggestion
    (let* ((results-key (seq-mapcat
                         (lambda (dir)
                           (directory-files
                            dir t
                            (format "\\`%s.*\\.\\(?:%s\\)\\'"
                                    (regexp-quote key)
                                    (mapconcat #'regexp-quote extensions "\\|"))))
                         dirs)))
```

Reply:
WDYT about adding a defcustom for this?

I'm agnostic ATM.

Reply:
I didn’t see a reason for a defcustom in the bibtex-completion version, so I wasn’t compelled to propose one here.

Reply:
I'm not sure either. Maybe if someone has keys that are prefixes of other keys this could cause false positives? I would be inclined to add a defcustom if/when someone asks for it.

More likely, people might have other library organization schemes (directory per reference?) We can also think about how to accommodate those if somebody asks.

More importantly, there's the performance concern with performing a directory scan to check if a reference has any files. This could really slow down the "has:note" testing when compiling completion candidates. Bibtex-completion in fact [doesn't seem to do this scan.](https://github.com/tmalsburg/helm-bibtex/blob/aa775340ba691d2322948bfdc6a88158568a1399/bibtex-completion.el#L101)

Reply:
> More importantly, there's the performance concern with performing a directory scan to check if a reference has any files. 

That was the main thing I was wondering about.

Reply:
I think the proper way to do this is to scan the relevant directories once (and add a file watch) and cache which references have files in them. Then use that cache when preparing the completion candidates. But this would need a bigger code change.

Reply:
Performing a potentially O(n^2) operation (in the number of entries) doesn't seem like a good idea. So we do need a defcustom to allow this, and we need to let-bind it to nil when compiling candidates.

Reply:
I would think at minimum that function needs another optional arg so it can be turned off so that the performance-sensitive operations involved in generating a candidate list don't get bogged down.

Reply:
The code to do that is almost the same as adding the defcustom though, since you need to handle both cases. And the defcustom does allow people to turn off the feature if they're getting false positives. Would you be willing to implement this @localauthor?

Reply:
When I have some time I’ll give it a shot. Thanks for the review and suggestions.

Reply:
I tried to incorporate your changes, @roshanshariff , but could only get it working without them. Any thoughts/suggestions?